### DISCUSSION TITLE
What is the Proper Way to Set All Table Metrics at Once?

### DISCUSSION THREAD

**Discussion by robloo** [fontforge/fontforge#4931] [Q&A]
I have a font being built with a FontForge python script. Everything is perfect for the first part of the script. Glyphs are imported then the initial ascent/descent font metrics are set correctly. However, after setting the OS/2 and hhea table metrics the values are incorrect.

It is as if FontForge is trying to recalculate after every single property is set instead of updating at the end. If these values are updated in the GUI (Font Information > General and Font Information > OS/2 > Metrics) all values are committed only after clicking OK and no adjustments are made -- values are correct. How do I get this same behavior from a python script? Can I commit multiple OS/2 hhea table changes at once without FontForge trying to adjust things after each value is set?

This works fine
```python
# Adjust each glyph's width to match the import default 1000x1000.
for glyph in font.glyphs():
    glyph.width = 1000

# Move the baseline from the default 200 to 0.
# This is done indirectly by setting both the ascent and descent.
# Remember FontForge will by default import with size 1000x1000 and 200 baseline.
font.ascent  = 1000 # From 800
font.descent = 0    # From 200

# Translate each glyph's position after moving the baseline.
translate_matrix = psMat.translate(0, 200)
for glyph in font.glyphs():
    glyph.transform(translate_matrix)

# Change the Em-Size of the font to match other symbol fonts.
# The convention is that TrueType fonts should have an Em-Size which is a power of 2.
# Setting this will scale the entire font (each glyph) to the new size.
font.em = 2048
```

However, I then attempt to set the OS/2 and hhea table metrics using the below code and it messes up the ascent/descent. I've tried many different orderings of these parameters.

```python
font.os2_use_typo_metrics = 0     # 0 means false

font.os2_typolinegap      = 0     # Type Line Gap
font.hhea_linegap         = 0     # HHead Line Gap

font.os2_windescent       = 0     # Win Descent
font.os2_typodescent      = 0     # Typo Descent
font.hhea_descent         = 0     # HHead Descent
font.os2_winascent        = 2048  # Win Ascent
font.os2_typoascent       = 2048  # Typo Ascent
font.hhea_ascent          = 2048  # HHead Ascent
font.ascent  = 2048
font.descent = 0
```

It results in the following:
![image](https://user-images.githubusercontent.com/17993847/157152801-4acb4afd-574a-4e55-a035-499a38db8e1c.png)
![image](https://user-images.githubusercontent.com/17993847/157152832-032854a7-88c2-4c6c-a6bf-6a21f7b16a84.png)

**Comment by robloo** [comment]
Is there anyone who is knowledgeable about Python scripting FontForge that can help with this? I'm not sure who to ask.

**Comment by skef** [comment]
This is a thorny issue and I'm not sure what to say, but you might get better help on the FontForge users mailing list: https://sourceforge.net/projects/fontforge/lists/fontforge-users

**Reply by robloo** [reply]
Thanks, I'll give that a try.
